agents/supervisor/supervisor.go

```diff
@@ -127,18 +134,21 @@ func (s *Supervisor) setAgentProcesses(agentProcesses map[string]*agentpb.SetSta
 	// stop first to avoid extra load
 	for _, agentID := range toStop {
 		agent := s.agentProcesses[agentID]
+
 		agent.cancel()
 		<-agent.done // wait before releasing port
 
 		if err := s.portsRegistry.Release(agent.listenPort); err != nil {
 			s.l.Errorf("Failed to release port: %s.", err)
 		}
 		delete(s.agentProcesses, agentID)
+		delete(s.agentStatuses, agentID)
```

Comment:
Sorry, but now we also should hold `s.arw`.
I wonder if we can trigger this race with a test.